File: src/Compilers/CSharp/Portable/Syntax/XDocumentExtensions.cs
// -----------------------------------------------------------------------
// <copyright file="XDocumentExtensions.cs" company="Altemiq">
// Copyright (c) Altemiq. All rights reserved.
// </copyright>
// -----------------------------------------------------------------------

namespace Microsoft.CodeAnalysis.CSharp.Syntax;

using System.Xml.Linq;

/// <summary>
/// The <see cref="System.Xml.Linq"/> extensions.
/// </summary>
public static class XDocumentExtensions
{
    /// <summary>
    /// Converts the container in to <see cref="XmlNodeSyntax"/> elements.
    /// </summary>
    /// <param name="container">The container.</param>
    /// <returns>The transformed nodes.</returns>
    public static IEnumerable<XmlNodeSyntax> ToXmlNodes(this XContainer container) => container.Nodes().ToXmlNodes();

    /// <summary>
    /// Converts the nodes in to <see cref="XmlNodeSyntax"/> elements.
    /// </summary>
    /// <param name="nodes">The nodes.</param>
    /// <returns>The transformed nodes.</returns>
    public static IEnumerable<XmlNodeSyntax> ToXmlNodes(this IEnumerable<XNode> nodes) => nodes.Select(node => node.ToXmlNode());

    /// <summary>
    /// Converts the document in to <see cref="XmlNodeSyntax"/> elements.
    /// </summary>
    /// <param name="document">The document.</param>
    /// <returns>The transformed nodes.</returns>
    public static XmlNodeSyntax ToXmlNode(this XDocument document) => document.Root is { } root
        ? root.ToXmlNode()
        : throw new InvalidOperationException();

    /// <summary>
    /// Converts the node into a <see cref="XmlNodeSyntax"/>.
    /// </summary>
    /// <param name="node">The node.</param>
    /// <returns>The transformed node.</returns>
    public static XmlNodeSyntax ToXmlNode(this XNode node)
    {
        return node switch
        {
            // Text
            XText text => SyntaxFactory.XmlText(text.Value),

            // documentation specific attributes
            XElement { Name.LocalName: Roslyn.Utilities.DocumentationCommentXmlNames.Elements.Parameter } element => SyntaxFactory.XmlParamElement(GetName(element), element.ToXmlNodes().ToArray()),
            XElement { Name.LocalName: Roslyn.Utilities.DocumentationCommentXmlNames.Elements.ParameterReference } element => SyntaxFactory.XmlParamRefElement(GetName(element)),
            XElement { Name.LocalName: Roslyn.Utilities.DocumentationCommentXmlNames.Elements.Summary } element => SyntaxFactory.XmlSummaryElement(element.ToXmlNodes().ToArray()),
            XElement { Name.LocalName: Roslyn.Utilities.DocumentationCommentXmlNames.Elements.Returns } element => SyntaxFactory.XmlReturnsElement(element.ToXmlNodes().ToArray()),
            XElement { Name.LocalName: Roslyn.Utilities.DocumentationCommentXmlNames.Elements.Remarks } element => SyntaxFactory.XmlRemarksElement(element.ToXmlNodes().ToArray()),
            XElement { Name.LocalName: Roslyn.Utilities.DocumentationCommentXmlNames.Elements.Para } element => SyntaxFactory.XmlParaElement(element.ToXmlNodes().ToArray()),
            XElement { Name.LocalName: Roslyn.Utilities.DocumentationCommentXmlNames.Elements.See } element => CreateSeeElement(element),
            XElement { Name.LocalName: Roslyn.Utilities.DocumentationCommentXmlNames.Elements.SeeAlso } element => CreateSeeAlsoElement(element),
            XElement { Name.LocalName: Roslyn.Utilities.DocumentationCommentXmlNames.Elements.Exception } element => SyntaxFactory.XmlExceptionElement(CreateCrefElement(element), element.ToXmlNodes().ToArray()),
            XElement { Name.LocalName: Roslyn.Utilities.DocumentationCommentXmlNames.Elements.Permission } element => SyntaxFactory.XmlPermissionElement(CreateCrefElement(element), element.ToXmlNodes().ToArray()),
            XElement { Name.LocalName: Roslyn.Utilities.DocumentationCommentXmlNames.Elements.Example } element => SyntaxFactory.XmlExampleElement(element.ToXmlNodes().ToArray()),
            XElement { Name.LocalName: Roslyn.Utilities.DocumentationCommentXmlNames.Elements.Value } element => SyntaxFactory.XmlValueElement(element.ToXmlNodes().ToArray()),
            XElement { Name.LocalName: Roslyn.Utilities.DocumentationCommentXmlNames.Elements.Placeholder } element => SyntaxFactory.XmlPlaceholderElement(element.ToXmlNodes().ToArray()),
            XElement { Name.LocalName: Roslyn.Utilities.DocumentationCommentXmlNames.Elements.ThreadSafety } => SyntaxFactory.XmlThreadSafetyElement(),

            // Generic
            XElement { Name.LocalName: Roslyn.Utilities.DocumentationCommentXmlNames.Elements.TypeParameter } element => SyntaxFactory.XmlElement(Roslyn.Utilities.DocumentationCommentXmlNames.Elements.TypeParameter, SyntaxFactory.List(element.ToXmlNodes())).AddStartTagAttributes(SyntaxFactory.XmlNameAttribute(GetName(element))),
            XElement { Name.LocalName: Roslyn.Utilities.DocumentationCommentXmlNames.Elements.TypeParameterReference } element => SyntaxFactory.XmlElement(Roslyn.Utilities.DocumentationCommentXmlNames.Elements.TypeParameterReference, SyntaxFactory.List(element.ToXmlNodes())).AddStartTagAttributes(SyntaxFactory.XmlNameAttribute(GetName(element))),

            // List
            XElement { Name.LocalName: Roslyn.Utilities.DocumentationCommentXmlNames.Elements.List } element => SyntaxFactory.XmlElement(Roslyn.Utilities.DocumentationCommentXmlNames.Elements.List, SyntaxFactory.List(element.ToXmlNodes())).AddStartTagAttributes(SyntaxFactory.XmlTextAttribute(Roslyn.Utilities.DocumentationCommentXmlNames.Attributes.Type, GetAttributeValue(element, Roslyn.Utilities.DocumentationCommentXmlNames.Attributes.Type))),
            XElement { Name.LocalName: Roslyn.Utilities.DocumentationCommentXmlNames.Elements.ListHeader } element => SyntaxFactory.XmlElement(Roslyn.Utilities.DocumentationCommentXmlNames.Elements.ListHeader, SyntaxFactory.List(element.ToXmlNodes())),
            XElement { Name.LocalName: Roslyn.Utilities.DocumentationCommentXmlNames.Elements.Item } element => SyntaxFactory.XmlElement(Roslyn.Utilities.DocumentationCommentXmlNames.Elements.Item, SyntaxFactory.List(element.ToXmlNodes())),
            XElement { Name.LocalName: Roslyn.Utilities.DocumentationCommentXmlNames.Elements.Term } element => SyntaxFactory.XmlElement(Roslyn.Utilities.DocumentationCommentXmlNames.Elements.Term, SyntaxFactory.List(element.ToXmlNodes())),
            XElement { Name.LocalName: Roslyn.Utilities.DocumentationCommentXmlNames.Elements.Description } element => SyntaxFactory.XmlElement(Roslyn.Utilities.DocumentationCommentXmlNames.Elements.Description, SyntaxFactory.List(element.ToXmlNodes())),

            // Just translate directly
            XElement { Name.LocalName: { } name, HasAttributes: true } element => SyntaxFactory.XmlElement(name, SyntaxFactory.List(element.ToXmlNodes())).AddStartTagAttributes([.. GetAttributes(element)]),
            XElement { Name.LocalName: { } name } element => SyntaxFactory.XmlElement(name, SyntaxFactory.List(element.ToXmlNodes())),

            _ => throw new NotSupportedException(),
        };

        static string GetName(XElement element)
        {
            return GetAttributeValue(element, Roslyn.Utilities.DocumentationCommentXmlNames.Attributes.Name);
        }

        static string GetAttributeValue(XElement element, string name)
        {
            return element.Attribute(name)!.Value;
        }

        static bool TryGetAttributeValue(XElement element, string name, out string value)
        {
            if (element.Attribute(name) is { } attribute)
            {
                value = attribute.Value;
                return true;
            }

            value = string.Empty;
            return false;
        }

        static IEnumerable<XmlAttributeSyntax> GetAttributes(XElement element)
        {
            return element.Attributes().Select<XAttribute, XmlAttributeSyntax>(attribute => attribute switch
            {
                { Name.LocalName: Roslyn.Utilities.DocumentationCommentXmlNames.Attributes.Name, Value: { } value } => SyntaxFactory.XmlNameAttribute(value),
                { Name.LocalName: Roslyn.Utilities.DocumentationCommentXmlNames.Attributes.Cref, Value: { } value } => SyntaxFactory.XmlCrefAttribute(GetCrefElement(value)),
                { Name.LocalName: { } name, Value: { } value } => SyntaxFactory.XmlTextAttribute(name, value),
            });
        }

        static XmlEmptyElementSyntax CreateSeeElement(XElement element)
        {
            if (TryGetAttributeValue(element, Roslyn.Utilities.DocumentationCommentXmlNames.Attributes.Cref, out var cref))
            {
                return SyntaxFactory.XmlSeeElement(GetCrefElement(cref));
            }

            if (TryGetAttributeValue(element, Roslyn.Utilities.DocumentationCommentXmlNames.Attributes.Langword, out var keyword))
            {
                return SyntaxFactory.XmlEmptyElement(Roslyn.Utilities.DocumentationCommentXmlNames.Elements.See).AddAttributes(
                    SyntaxFactory.XmlTextAttribute(Roslyn.Utilities.DocumentationCommentXmlNames.Attributes.Langword, keyword));
            }

            if (TryGetAttributeValue(element, Roslyn.Utilities.DocumentationCommentXmlNames.Attributes.Href, out var href))
            {
                return SyntaxFactory.XmlEmptyElement(Roslyn.Utilities.DocumentationCommentXmlNames.Elements.See).AddAttributes(
                    SyntaxFactory.XmlTextAttribute(Roslyn.Utilities.DocumentationCommentXmlNames.Attributes.Href, href));
            }

            throw new InvalidOperationException();
        }

        static XmlEmptyElementSyntax CreateSeeAlsoElement(XElement element)
        {
            if (TryGetAttributeValue(element, Roslyn.Utilities.DocumentationCommentXmlNames.Attributes.Cref, out var cref))
            {
                return SyntaxFactory.XmlSeeAlsoElement(GetCrefElement(cref));
            }

            if (TryGetAttributeValue(element, Roslyn.Utilities.DocumentationCommentXmlNames.Attributes.Href, out var href))
            {
                return SyntaxFactory.XmlEmptyElement(Roslyn.Utilities.DocumentationCommentXmlNames.Elements.See).AddAttributes(
                    SyntaxFactory.XmlTextAttribute(Roslyn.Utilities.DocumentationCommentXmlNames.Attributes.Href, href));
            }

            throw new InvalidOperationException();
        }

        static CrefSyntax CreateCrefElement(XElement element)
        {
            return TryGetAttributeValue(element, Roslyn.Utilities.DocumentationCommentXmlNames.Attributes.Cref, out var cref)
                ? GetCrefElement(cref)
                : throw new InvalidOperationException();
        }

        static CrefSyntax GetCrefElement(string cref)
        {
            var colonIndex = cref.IndexOf(':');
            if (colonIndex < 0)
            {
                return SyntaxFactory.NameMemberCref(SyntaxFactory.QualifiedName(cref));
            }

            var type = cref.Substring(0, colonIndex);
            colonIndex++;
            return type switch
            {
                "T" => SyntaxFactory.TypeCref(SyntaxFactory.QualifiedName(cref.Substring(colonIndex))),
                _ => SyntaxFactory.NameMemberCref(SyntaxFactory.QualifiedName(cref.Substring(colonIndex))),
            };
        }
    }
}

File: src/Compilers/VisualBasic/Portable/Syntax/XDocumentExtensions.vb
' -----------------------------------------------------------------------
' <copyright file="XDocumentExtensions.vb" company="Altemiq">
' Copyright (c) Altemiq. All rights reserved.
' </copyright>
' -----------------------------------------------------------------------

Imports System.Runtime.CompilerServices

Namespace Syntax
    ''' <summary>
    ''' The <see cref="System.Xml.Linq"/> extensions.
    ''' </summary>
    Public Module XDocumentExtensions
        ''' <summary>
        ''' Converts the container in to <see cref="XmlNodeSyntax"/> elements.
        ''' </summary>
        ''' <param name="container">The container.</param>
        ''' <returns>The transformed nodes.</returns>
        <Extension>
        Public Function ToXmlNodes(container As XContainer) As IEnumerable(Of XmlNodeSyntax)
            Return container.Nodes.ToXmlNodes
        End Function

        ''' <summary>
        ''' Converts the nodes in to <see cref="XmlNodeSyntax"/> elements.
        ''' </summary>
        ''' <param name="nodes">The nodes.</param>
        ''' <returns>The transformed nodes.</returns>
        <Extension>
        Public Function ToXmlNodes(nodes As IEnumerable(Of XNode)) As IEnumerable(Of XmlNodeSyntax)
            Return nodes.Select(Function(node) node.ToXmlNode)
        End Function

        ''' <summary>
        ''' Converts the document in to <see cref="XmlNodeSyntax"/> elements.
        ''' </summary>
        ''' <param name="document">The document.</param>
        ''' <returns>The transformed nodes.</returns>
        <Extension>
        Public Function ToXmlNode(document As XDocument) As XmlNodeSyntax
            If document.Root IsNot Nothing Then
                Return document.Root.ToXmlNode
            End If

            Throw New InvalidOperationException
        End Function

        ''' <summary>
        ''' Converts the node into a <see cref="XmlNodeSyntax"/>.
        ''' </summary>
        ''' <param name="node">The node.</param>
        ''' <returns>The transformed node.</returns>
        <Extension>
        Public Function ToXmlNode(node As XNode) As XmlNodeSyntax
            Dim text = TryCast(node, XText)
            If text IsNot Nothing Then
                Return SyntaxFactory.XmlText(text.Value)
            End If

            Dim element = TryCast(node, XElement)
            If element IsNot Nothing Then
                Dim name As String = element.Name.LocalName

                ' documentation specific attributes
                If name = Roslyn.Utilities.DocumentationCommentXmlNames.Elements.Parameter Then
                    Return SyntaxFactory.XmlParamElement(GetName(element), element.ToXmlNodes.ToArray)
                End If
                If name = Roslyn.Utilities.DocumentationCommentXmlNames.Elements.ParameterReference Then
                    Return SyntaxFactory.XmlParamRefElement(GetName(element))
                End If
                If name = Roslyn.Utilities.DocumentationCommentXmlNames.Elements.Summary Then
                    Return SyntaxFactory.XmlSummaryElement(element.ToXmlNodes.ToArray)
                End If
                If name = Roslyn.Utilities.DocumentationCommentXmlNames.Elements.Returns Then
                    Return SyntaxFactory.XmlReturnsElement(element.ToXmlNodes.ToArray)
                End If
                If name = Roslyn.Utilities.DocumentationCommentXmlNames.Elements.Remarks Then
                    Return SyntaxFactory.XmlRemarksElement(element.ToXmlNodes.ToArray)
                End If
                If name = Roslyn.Utilities.DocumentationCommentXmlNames.Elements.Para Then
                    Return SyntaxFactory.XmlParaElement(element.ToXmlNodes.ToArray)
                End If
                If name = Roslyn.Utilities.DocumentationCommentXmlNames.Elements.See Then
                    Return CreateSeeElement(element)
                End If
                If name = Roslyn.Utilities.DocumentationCommentXmlNames.Elements.SeeAlso Then
                    Return CreateSeeAlsoElement(element)
                End If
                If name = Roslyn.Utilities.DocumentationCommentXmlNames.Elements.Exception Then
                    Return SyntaxFactory.XmlExceptionElement(CreateCrefElement(element), element.ToXmlNodes.ToArray)
                End If
                If name = Roslyn.Utilities.DocumentationCommentXmlNames.Elements.Permission Then
                    Return SyntaxFactory.XmlPermissionElement(CreateCrefElement(element), element.ToXmlNodes.ToArray)
                End If
                If name = Roslyn.Utilities.DocumentationCommentXmlNames.Elements.Example Then
                    Return SyntaxFactory.XmlExampleElement(element.ToXmlNodes.ToArray)
                End If
                If name = Roslyn.Utilities.DocumentationCommentXmlNames.Elements.Value Then
                    Return SyntaxFactory.XmlValueElement(element.ToXmlNodes.ToArray)
                End If
                If name = Roslyn.Utilities.DocumentationCommentXmlNames.Elements.Placeholder Then
                    Return SyntaxFactory.XmlPlaceholderElement(element.ToXmlNodes.ToArray)
                End If
                If name = Roslyn.Utilities.DocumentationCommentXmlNames.Elements.ThreadSafety Then
                    Return SyntaxFactory.XmlThreadSafetyElement
                End If

                ' Generic
                If name = Roslyn.Utilities.DocumentationCommentXmlNames.Elements.TypeParameter Then
                    Return SyntaxFactory.XmlElement(
                        Roslyn.Utilities.DocumentationCommentXmlNames.Elements.TypeParameter,
                        SyntaxFactory.List(element.ToXmlNodes)).AddStartTagAttributes(SyntaxFactory.XmlNameAttribute(GetName(element)))
                End If
                If name = Roslyn.Utilities.DocumentationCommentXmlNames.Elements.TypeParameterReference Then
                    Return SyntaxFactory.XmlElement(
                        Roslyn.Utilities.DocumentationCommentXmlNames.Elements.TypeParameterReference,
                        SyntaxFactory.List(element.ToXmlNodes)).AddStartTagAttributes(SyntaxFactory.XmlNameAttribute(GetName(element)))
                End If

                ' List
                If name = Roslyn.Utilities.DocumentationCommentXmlNames.Elements.List Then
                    Return SyntaxFactory.XmlElement(
                        Roslyn.Utilities.DocumentationCommentXmlNames.Elements.List,
                        SyntaxFactory.List(element.ToXmlNodes)).AddStartTagAttributes(
                            CreateAttribute(
                                Roslyn.Utilities.DocumentationCommentXmlNames.Attributes.Type,
                                GetAttributeValue(element, Roslyn.Utilities.DocumentationCommentXmlNames.Attributes.Type)))
                End If
                If name = Roslyn.Utilities.DocumentationCommentXmlNames.Elements.ListHeader Then
                    Return SyntaxFactory.XmlElement(
                        Roslyn.Utilities.DocumentationCommentXmlNames.Elements.ListHeader,
                        SyntaxFactory.List(element.ToXmlNodes))
                End If
                If name = Roslyn.Utilities.DocumentationCommentXmlNames.Elements.Item Then
                    Return SyntaxFactory.XmlElement(
                        Roslyn.Utilities.DocumentationCommentXmlNames.Elements.Item,
                        SyntaxFactory.List(element.ToXmlNodes))
                End If
                If name = Roslyn.Utilities.DocumentationCommentXmlNames.Elements.Term Then
                    Return SyntaxFactory.XmlElement(
                        Roslyn.Utilities.DocumentationCommentXmlNames.Elements.Term,
                        SyntaxFactory.List(element.ToXmlNodes))
                End If
                If name = Roslyn.Utilities.DocumentationCommentXmlNames.Elements.Description Then
                    Return SyntaxFactory.XmlElement(
                        Roslyn.Utilities.DocumentationCommentXmlNames.Elements.Description,
                        SyntaxFactory.List(element.ToXmlNodes))
                End If

                ' Just translate directly
                If element.HasAttributes Then
                    Return SyntaxFactory.XmlElement(name, SyntaxFactory.List(element.ToXmlNodes)).AddStartTagAttributes(
                        GetAttributes(element).ToArray())
                End If

                Return SyntaxFactory.XmlElement(name, SyntaxFactory.List(element.ToXmlNodes))
            End If

            Throw New NotSupportedException
        End Function

        Private Function GetName(element As XElement) As String
            Return GetAttributeValue(element, Roslyn.Utilities.DocumentationCommentXmlNames.Attributes.Name)
        End Function

        Private Function GetAttributeValue(element As XElement, name As String) As String
            Return element.Attribute(name).Value
        End Function

        Private Function TryGetAttributeValue(element As XElement, name As String, ByRef value As String) As Boolean
            Dim attribute = element.Attribute(name)
            If attribute IsNot Nothing Then
                value = attribute.Value
                Return True
            End If

            value = String.Empty
            Return False
        End Function

        Private Function GetAttributes(element As XElement) As IEnumerable(Of BaseXmlAttributeSyntax)
            Return element.Attributes.Select(Function(attribute As XAttribute) As BaseXmlAttributeSyntax
                                                 If attribute.Value IsNot Nothing Then
                                                     If attribute.Name.LocalName = Roslyn.Utilities.DocumentationCommentXmlNames.Attributes.Name Then
                                                         Return SyntaxFactory.XmlNameAttribute(attribute.Value)
                                                     End If

                                                     If attribute.Name.LocalName = Roslyn.Utilities.DocumentationCommentXmlNames.Attributes.Cref Then
                                                         Return SyntaxFactory.XmlCrefAttribute(GetCrefElement(attribute.Value))
                                                     End If

                                                     Return CreateAttribute(
                                                         attribute.Name.LocalName,
                                                         attribute.Value)
                                                 End If

                                                 Throw New InvalidOperationException
                                             End Function)
        End Function

        Private Function CreateAttribute(name As String, value As String) As XmlAttributeSyntax
            Return SyntaxFactory.XmlAttribute(
                SyntaxFactory.XmlName(
                    Nothing,
                    SyntaxFactory.XmlNameToken(name, SyntaxKind.XmlNameToken)),
                SyntaxFactory.XmlString(
                    SyntaxFactory.Token(SyntaxKind.DoubleQuoteToken),
                    SyntaxFactory.TokenList(SyntaxFactory.XmlTextLiteral(value)),
                    SyntaxFactory.Token(SyntaxKind.DoubleQuoteToken)).AddTextTokens()).
                WithLeadingTrivia(SyntaxFactory.Whitespace(" "))
        End Function

        Private Function CreateSeeElement(element As XElement) As XmlEmptyElementSyntax
            Dim value As String = Nothing
            If TryGetAttributeValue(element, Roslyn.Utilities.DocumentationCommentXmlNames.Attributes.Cref, value) Then
                Return SyntaxFactory.XmlSeeElement(GetCrefElement(value))
            End If

            If TryGetAttributeValue(element, Roslyn.Utilities.DocumentationCommentXmlNames.Attributes.Langword, value) Then
                Return SyntaxFactory.XmlEmptyElement(Roslyn.Utilities.DocumentationCommentXmlNames.Elements.See).AddAttributes(
                    CreateAttribute(Roslyn.Utilities.DocumentationCommentXmlNames.Attributes.Langword, GetLangword(value)))
            End If

            If TryGetAttributeValue(element, Roslyn.Utilities.DocumentationCommentXmlNames.Attributes.Href, value) Then
                Return SyntaxFactory.XmlEmptyElement(Roslyn.Utilities.DocumentationCommentXmlNames.Elements.See).AddAttributes(
                    CreateAttribute(Roslyn.Utilities.DocumentationCommentXmlNames.Attributes.Href, value))
            End If

            Throw New InvalidOperationException
        End Function

        Private Function CreateSeeAlsoElement(element As XElement) As XmlEmptyElementSyntax
            Dim value As String = Nothing
            If TryGetAttributeValue(element, Roslyn.Utilities.DocumentationCommentXmlNames.Attributes.Cref, value) Then
                Return SyntaxFactory.XmlSeeAlsoElement(GetCrefElement(value))
            End If

            If TryGetAttributeValue(element, Roslyn.Utilities.DocumentationCommentXmlNames.Attributes.Href, value) Then
                Return SyntaxFactory.XmlEmptyElement(Roslyn.Utilities.DocumentationCommentXmlNames.Elements.See).AddAttributes(
                    CreateAttribute(
                        Roslyn.Utilities.DocumentationCommentXmlNames.Attributes.Href,
                        value))
            End If

            Throw New InvalidOperationException
        End Function

        Private Function CreateCrefElement(element As XElement) As CrefReferenceSyntax
            Dim value As String = Nothing
            If TryGetAttributeValue(element, Roslyn.Utilities.DocumentationCommentXmlNames.Attributes.Cref, value) Then
                Return GetCrefElement(value)
            End If
            Throw New InvalidOperationException
        End Function

        Private Function GetCrefElement(cref As String) As CrefReferenceSyntax
            Dim colonIndex = cref.IndexOf(":"c)
            If colonIndex < 0 Then
                Return SyntaxFactory.CrefReference(QualifiedName(cref))
            End If

            colonIndex += 1
            Return SyntaxFactory.CrefReference(QualifiedName(cref.Substring(colonIndex)))
        End Function

        Private Function GetLangword(value As String) As String
            If value = "null" Then
                Return "Nothing"
            End If

            If value = "true" Then
                Return Boolean.TrueString
            End If

            If value = "false" Then
                Return Boolean.FalseString
            End If

            If value = "static" Then
                Return "Shared"
            End If

            If value = "this" Then
                Return "Me"
            End If

            If value = "base" Then
                Return "MyBase"
            End If

            If value = "overload" Then
                Return "Overloads"
            End If

            If value = "override" Then
                Return "Overrides"
            End If

            Return value
        End Function

        <Extension>
        Private Function AddStartTagAttributes(element As XmlElementSyntax, ParamArray items() As BaseXmlAttributeSyntax) As XmlElementSyntax
            Return element.WithStartTag(element.StartTag.WithAttributes(element.StartTag.Attributes.AddRange(items)))
        End Function
    End Module
End Namespace